src/webapi/rendering_context.rs

```diff
@@ -27,7 +183,771 @@ impl RenderingContext for CanvasRenderingContext2d {
+        ).unwrap()
+    }
+
+    /// Creates a new, blank ImageData object with the specified dimensions. 
+    /// All of the pixels in the new object are transparent black.
+    /// 
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/createImageData)
+    // https://html.spec.whatwg.org/#2dcontext:dom-context-2d-createimagedata
+    pub fn create_image_data_id(&self, image_data: ImageData) -> Result<ImageData, IndexSizeError> {
+        js_try! (
+            return @{&self.0}.createImageData(@{image_data});
+        ).unwrap()
+    }
+
+    /// Creates a pattern using the specified image (a CanvasImageSource). It repeats the source in 
+    /// the directions specified by the repetition argument. This method returns a CanvasPattern.
+    /// 
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/createPattern)
+    // https://html.spec.whatwg.org/#2dcontext:dom-context-2d-createpattern
+    pub fn create_pattern_image(&self, image: ImageElement, repetition: Option<Repetition>) -> CanvasPattern {
+        let repetition_string = match repetition {
+            Some(Repetition::Repeat) | None => {
+                "repeat"
+            }
+
+            Some(Repetition::RepeatX) => {
+                "repeat-x"
+            }
+
+            Some(Repetition::RepeatY) => {
+                "repeat-y"
+            }
+
+            Some(Repetition::NoRepeat) => {
+                "no-repeat"
+            }
+        };
+
+        js! (
+            return @{&self.0}.createPattern(@{image}, @{repetition_string});
+        ).try_into().unwrap()
+    }
+
+    /// Creates a radial gradient given by the coordinates of the two circles represented by the parameters. 
+    /// This method returns a CanvasGradient.
+    /// 
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/createRadialGradient)
+    // https://html.spec.whatwg.org/#2dcontext:dom-context-2d-createradialgradient
+    pub fn create_radial_gradient(&self, x0: f64, y0: f64, r0: f64, x1: f64, y1: f64, r1: f64) -> CanvasGradient {
+        js! (
+            return @{&self.0}.createRadialGradient(@{x0}, @{y0}, @{r0}, @{x1}, @{y1}, @{r1});
+        ).try_into().unwrap()
+    }
+
+    /// Draws a focus ring around the current path or given path, If a given element is focused.
+    /// 
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/drawFocusIfNeeded)
+    // https://html.spec.whatwg.org/#2dcontext:dom-context-2d-drawfocusifneeded
+    pub fn draw_focus_if_needed< T: IHtmlElement >(&self, element: &T) {
+        js! { @(no_return)
+            @{&self.0}.drawFocusIfNeeded(@{element.as_ref()});
+        }
+    }
+
+    /// Provides different ways to draw an image onto the canvas.
+    /// TODO: Potentially throw more than just TypeMismatchError
```

Comment:
`TODO`s should probably be left in normal `//` comments; everything which starts with `///` will end up in the user-visible documentation. (: